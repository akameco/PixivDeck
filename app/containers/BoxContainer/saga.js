// @flow
import { shell } from 'electron'
import type { Saga } from 'redux-saga'
import { call, takeEvery } from 'redux-saga/effects'
import * as Actions from './constants'

function* openPixiv({ id }: { id: number }): Saga<void> {
  yield call(
    shell.openExternal,
    `http://www.pixiv.net/member_illust.php?mode=medium&illust_id=${id}`
  )
}

export default function* root(): Saga<void> {
  yield takeEvery(Actions.OPEN_PIXIV, openPixiv)
}
