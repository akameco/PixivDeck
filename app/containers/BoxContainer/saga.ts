import { shell } from 'electron'
import { call, takeEvery } from 'redux-saga/effects'
import * as Actions from './constants'

function* openPixiv({ id }: { id: number }) {
  yield call(
    shell.openExternal,
    `http://www.pixiv.net/member_illust.php?mode=medium&illust_id=${id}`
  )
}

export default function* root() {
  yield takeEvery(Actions.OPEN_PIXIV, openPixiv)
}
