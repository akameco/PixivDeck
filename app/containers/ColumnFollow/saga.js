// @flow
import type { Saga } from 'redux-saga'
import { difference } from 'lodash'
import { put, select, call, takeEvery, delay } from 'redux-saga/effects'
import { addTable } from 'containers/ColumnManager/actions'
import { getMyId } from 'containers/LoginModal/selectors'
import { addNotifyWithIllust } from '../Notify/actions'
import { FOLLOW_SUCCESS } from '../FollowButton/constants'
import * as fetchColumn from '../Column/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { ColumnId } from './reducer'
import * as selectors from './selectors'

type Action = { id: ColumnId }

export function* addFollowColumn({ id }: Action): Saga<void> {
  const ids: Array<?ColumnId> = yield select(selectors.makeSelectIds())
  if (ids.every(v => v !== id)) {
    yield put(actions.addColumnSuccess(id))
  }

  yield put(addTable(`follow-${id}`, { columnId: id, type: 'FOLLOW' }))
}

const getEndpoint = (userId, restrict) =>
  `/v2/illust/follow?user_id=${userId}&restrict=${restrict}`

function* fetchFollow({ id }: Action): Saga<void> {
  try {
    const { ids, nextUrl } = yield select(selectors.makeSelectColumn(), { id })
    const endpoint = nextUrl ? nextUrl : getEndpoint(yield select(getMyId), id)
    yield call(fetchColumn.fetchColumn, endpoint, id, actions, ids)
  } catch (error) {
    yield put(actions.fetchFailre(id, error))
  }
}

function* fetchNew({ id }: Action): Saga<void> {
  const { ids } = yield select(selectors.makeSelectColumn(), { id })
  const endpoint = getEndpoint(yield select(getMyId), id)
  yield call(fetchColumn.fetchNew, { endpoint, id, ids, order: true }, actions)
}

// TODO キャンセル
function* fetchNewWatch(action: Action) {
  try {
    while (true) {
      const { interval } = yield select(selectors.makeSelectColumn(), action)
      // 増加したイラストをチェック
      const { ids } = yield select(selectors.makeSelectColumn(), action)
      yield call(fetchNew, action)
      const { ids: nextIds } = yield select(
        selectors.makeSelectColumn(),
        action
      )

      if (ids.length > 0) {
        const diffIllusts = difference(nextIds, ids)
        for (const id of diffIllusts) {
          yield put(addNotifyWithIllust('新着イラスト', id))
        }
      }

      yield delay(interval)
    }
  } catch (error) {
    // TODO エラーハンドリング
  }
}

function* handleFollowSuccess({ restrict }): Saga<*> {
  const ids: Array<?ColumnId> = yield select(selectors.makeSelectIds())
  if (ids.some(v => v === restrict)) {
    yield call(fetchFollow, { id: restrict })
  }
}

export default function* root(): Saga<void> {
  yield takeEvery(Actions.ADD_COLUMN, addFollowColumn)
  yield takeEvery([Actions.FETCH, Actions.FETCH_NEXT], fetchFollow)

  yield takeEvery(Actions.ADD_COLUMN_SUCCESS, fetchNewWatch)
  yield takeEvery(FOLLOW_SUCCESS, handleFollowSuccess)
}
