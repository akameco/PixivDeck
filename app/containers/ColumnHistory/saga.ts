import { addTable } from 'containers/ColumnManager/actions'
import { put, takeEvery } from 'redux-saga/effects'
import { OPEN_ILLUST_VIEWER } from '../IllustPreview/constants'
import { OPEN_MANGA_PREVIEW } from '../MangaPreview/constants'
import * as Actions from './constants'
import * as actions from './actions'

export function* addHistoryColumn() {
  yield put(
    addTable('HISTORY', {
      columnId: 'history',
      type: 'HISTORY',
    })
  )
}
interface Action {
  id: number
}
export function* addHistory({ id }: Action) {
  yield put(actions.addHistory(id))
}
export default function* root() {
  yield takeEvery(Actions.ADD_COLUMN_HISTORY, addHistoryColumn)
  yield takeEvery([OPEN_ILLUST_VIEWER, OPEN_MANGA_PREVIEW], addHistory)
}
