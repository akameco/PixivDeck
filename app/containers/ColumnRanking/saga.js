// @flow
import { type Saga } from 'redux-saga'
import {
  put,
  select,
  call,
  takeEvery,
  fork,
  take,
  delay,
} from 'redux-saga/effects'
import { addTable } from 'containers/ColumnManager/actions'
import * as api from '../Column/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { Mode } from './reducer'
import * as selectors from './selectors'

type Action = { id: Mode }

export function* addColumn({ id }: Action): Saga<void> {
  const modes: Array<?Mode> = yield select(selectors.makeSelectModes())
  if (modes.every(v => v !== id)) {
    yield put(actions.addColumnSuccess(id))
  }

  yield put(addTable(`ranking-${id}`, { columnId: id, type: 'RANKING' }))
}

const getEndpoint = id => `/v1/illust/ranking?mode=${id}`

export function* fetch(action: Action): Saga<void> {
  const { ids, nextUrl } = yield select(selectors.makeSelectColumn(), action)
  const endpoint = nextUrl ? nextUrl : getEndpoint(action.id)
  yield call(api.fetchColumn, endpoint, action.id, actions, ids)
}

export function* fetchNew({ id }: Action): Saga<void> {
  const { ids } = yield select(selectors.makeSelectColumn(), { id })
  yield call(
    api.fetchNew,
    { endpoint: getEndpoint(id), id, ids, order: 'overwrite' },
    actions
  )
}

// $FlowFixMe
export function* watchNewIllust(): Saga<*> {
  while (true) {
    const { id } = yield take(Actions.START_WATCH)
    const interval = yield select(selectors.getInterval, { id })
    yield delay(interval)
    yield call(fetchNew, { id })
    yield put(actions.watchNew(id))
  }
}

export default function* root(): Saga<void> {
  yield takeEvery(Actions.ADD_COLUMN, addColumn)
  yield takeEvery(Actions.FETCH, fetch)
  yield fork(watchNewIllust)
}
