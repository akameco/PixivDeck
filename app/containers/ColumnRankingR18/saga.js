// @flow
import { delay, type Saga } from 'redux-saga'
import { put, select, call, takeEvery, fork, take } from 'redux-saga/effects'
import { addTable } from 'containers/ColumnManager/actions'
import * as api from '../Column/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { R18Mode } from './reducer'
import * as selectors from './selectors'

type Action = { id: R18Mode }

export function* addColumn({ id }: Action): Saga<void> {
  const modes: Array<?R18Mode> = yield select(selectors.makeSelectModes())
  if (modes.every(v => v !== id)) {
    yield put(actions.addColumnSuccess(id))
  }

  yield put(
    addTable(`ranking-r18-${id}`, { columnId: id, type: 'RANKING_R18' })
  )
}

const getEndpoint = id => `/v1/illust/ranking?mode=${id}`

export function* fetch(action: Action): Saga<void> {
  const { ids, nextUrl } = yield select(selectors.makeSelectColumn(), action)
  const endpoint = nextUrl ? nextUrl : getEndpoint(action.id)
  yield call(api.fetchColumn, endpoint, action.id, actions, ids)
}

export function* fetchNew({ id }: Action): Saga<void> {
  const { ids } = yield select(selectors.makeSelectColumn(), { id })
  yield call(
    api.fetchNew,
    { endpoint: getEndpoint(id), id, ids, order: 'overwrite' },
    actions
  )
}

// $FlowFixMe
export function* watchNewIllust(): Saga<*> {
  while (true) {
    const { id } = yield take(Actions.START_WATCH)
    const interval = yield select(selectors.getInterval, { id })
    yield delay(interval)
    yield call(fetchNew, { id })
    yield put(actions.watchNew(id))
  }
}

export default function* root(): Saga<void> {
  yield takeEvery(Actions.ADD_COLUMN, addColumn)
  yield takeEvery(Actions.FETCH, fetch)
  yield fork(watchNewIllust)
}
