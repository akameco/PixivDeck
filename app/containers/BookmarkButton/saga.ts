import { call, put, takeEvery } from 'redux-saga/effects'
import * as api from 'containers/Api/sagas'
import { get } from '../Api/sagas'
import * as columnActions from '../ColumnBookmark/actions'
import * as Actions from './constants'
import * as actions from './actions'
import { Restrict } from './types'

interface Props {
  id: number
  restrict: Restrict
}
export function* bookmark({ id, restrict }: Props) {
  try {
    yield call(api.post, '/v2/illust/bookmark/add', {
      illustId: id,
      restrict,
    })
    yield put(actions.addBookmarkSuccess(id, restrict))
  } catch (error) {
    yield put(actions.addBookmarkFailer(id, error))
  }
}
export function* deleteTask({ id }: Props) {
  try {
    yield call(api.post, '/v1/illust/bookmark/delete', {
      illustId: id,
    })
    yield put(actions.deleteBookmarkSuccess(id))
    yield put(columnActions.removeItem('public', id))
  } catch (error) {
    yield put(actions.deleteBookmarkFailer(id, error))
  }
}

function* success({ id }: Props) {
  try {
    yield call(get, `/v1/illust/detail?illust_id=${id}`, true)
  } catch (error) {}
}

function* root() {
  yield takeEvery(Actions.ADD_BOOKMARK_REQUEST, bookmark)
  yield takeEvery(Actions.DELETE_BOOKMARK_REQUEST, deleteTask)
  yield takeEvery(
    [Actions.ADD_BOOKMARK_SUCCESS, Actions.DELETE_BOOKMARK_SUCCESS],
    success
  )
}

export default root
