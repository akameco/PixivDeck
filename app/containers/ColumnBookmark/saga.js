// @flow
import type { Saga } from 'redux-saga'
import { put, select, call, takeEvery } from 'redux-saga/effects'
import { addTable } from 'containers/ColumnManager/actions'
import { getMyId } from 'containers/LoginModal/selectors'
import { ADD_BOOKMARK_SUCCESS } from '../BookmarkButton/constants'
import * as column from '../Column/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { Action } from './actionTypes'
import type { ColumnId } from './reducer'
import { makeSelectColumn, makeSelectIds } from './selectors'

export function* addColumn({ id }: { id: 'public' }): Saga<void> {
  const ids: Array<?ColumnId> = yield select(makeSelectIds())
  if (ids.every(v => v !== id)) {
    yield put(actions.addColumnSuccess(id))
  }

  yield put(addTable(`bookmark-${id}`, { columnId: id, type: 'BOOKMARK' }))
}

const getEndpoint = (userId, restrict) =>
  `/v1/user/bookmarks/illust?user_id=${userId}&restrict=${restrict}`

export function* fetchBookmark({ id }: { id: 'public' }): Saga<void> {
  const { ids, nextUrl } = yield select(makeSelectColumn(), { id })
  const endpoint = nextUrl ? nextUrl : getEndpoint(yield select(getMyId), id)
  yield call(column.fetchColumn, endpoint, id, { ...actions }, ids)
}

export function* fetchNew({
  restrict: id,
}: {
  restrict: $PropertyType<Action, 'id'>,
}): Saga<void> {
  try {
    const { ids } = yield select(makeSelectColumn(), { id })
    const endpoint = getEndpoint(yield select(getMyId), id)
    yield call(column.fetchNew, { endpoint, id, ids, order: true }, actions)
  } catch (error) {
    yield put(actions.fetchFailre(id, error))
  }
}

export default function* root(): Saga<void> {
  yield takeEvery(Actions.ADD_COLUMN, addColumn)
  yield takeEvery([Actions.FETCH, Actions.FETCH_NEXT], fetchBookmark)
  yield takeEvery(ADD_BOOKMARK_SUCCESS, fetchNew)
}
