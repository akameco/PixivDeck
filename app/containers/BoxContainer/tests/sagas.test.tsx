import * as sagas from '../saga'

test('root', () => {
  const gen = sagas.default()
  const next = gen.next()
  expect(next.value).toMatchSnapshot()
})
