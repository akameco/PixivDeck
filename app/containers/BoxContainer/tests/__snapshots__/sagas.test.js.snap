// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`root 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "BoxContainer/OPEN_PIXIV",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "FORK",
}
`;
